src/sessionization.py

```diff
@@ -0,0 +1,148 @@
+def process_logfile(inputfile, outfile, ttl):
+    try:
+        fo=open(outfile, "w")
+
+        with open(inputfile) as f:
+            # get first line, which should be
+            # ip,date,time,zone,cik,accession,extention,code,size,idx,norefer,noagent,find,crawler,browser
+            # but we will make sure
+            fieldline=f.readline().split(",")
+ 
+            # check fields here: we only care about ip, date, time
+            fieldindices=[]
+            for field in fields:
+                    if field in fieldline:
+                        fieldindices.append(fieldline.index(field))
+            if (len(fieldindices) != 3):
+                print "Didn't find all fields, exiting!"
+                sys.exit()
+
+            
+            if DEBUG: print "field indices [ip, date, time] : ", fieldindices
+            
+
+            for l in f: # this handles large files better than readlines()
+                l2=l.split(",")
+
+                newtime=datetime.datetime.strptime(l2[fieldindices[1]] + " " + l2[fieldindices[2]], "%Y-%m-%d %H:%M:%S")
+                # update current time, we want to remove expired sessions
+                if 'curtime' not in vars():
+                    timediff=0
+                else:
+                    timediff=int((newtime-curtime).total_seconds())
+                curtime=newtime
+
+                # get IP of entry
+                lip=l2[fieldindices[0]]
+
+                # add state to the dictionary as a key if it doesn't exist, or increment value of existing key if it does
+                if lip in users:
+                    # IP already exists, need to increment pages and update TTL
+                    users[lip][1]+=1
+                    users[lip][2]=ttl
+                else:
+                    # instantiate new dictionary entry, [start time, number of pages, time to live]
+                    users[lip]=[curtime, 1, ttl]
+
+                # if time hasn't advanced, save us from iterating through all fields
+                if timediff == 0:
+                    continue
+
+                # now go through and check all remaining fields
+                for i in users.keys():
+                    if i==lip: # already updated this one, skip
+                        continue
+                    else:
+                        users[i][2]-=timediff
+                        if users[i][2] <= 0: # this session's up
+                            if DEBUG:
+                                print("Ending session")
+                                print users[i]
+                                
+                            # the last time is equal to current time adjusted by TTL field and TTL
+                            endtime=curtime-datetime.timedelta(seconds=ttl-users[i][2])
+                            fo.write("%s,%s,%s,%d,%d\n" %(i, str(users[i][0]), str(endtime), int((endtime-users[i][0]).total_seconds())+1, users[i][1]))
```

Comment:
Stuff Like this could be rolled into a utility function that would convert a closed session into a row string. Abstract that visual complexity would be great, especially because you do the exact same thing below when ending the file